qiskit/aqua/algorithms/phase_estimators/phase_estimator.py

```diff
@@ -0,0 +1,199 @@
+    amplitude that decreases with distance from the bitstring most closely approximating the phase.
+
+    **Reference:**
+
+    [1]: Michael A. Nielsen and Isaac L. Chuang. 2011.
+         Quantum Computation and Quantum Information: 10th Anniversary Edition (10th ed.).
+         Cambridge University Press, New York, NY, USA.
+
+    """
+
+    def __init__(self,
+                 num_evaluation_qubits: int,
+                 unitary: Optional[QuantumCircuit] = None,
+                 state_preparation: Optional[QuantumCircuit] = None,
+                 pe_circuit: Optional[QuantumCircuit] = None,
+                 num_unitary_qubits: Optional[int] = None,
+                 quantum_instance: Optional[Union[QuantumInstance, BaseBackend]] = None) -> None:
+
+        """Args:
+            num_evaluation_qubits: The number of qubits used in estimating the phase. The phase will
+                                   be estimated as a binary string with this many bits.
+            unitary: The circuit representing the unitary operator whose eigenvalues (via phase)
+                     will be measured. Exactly one of `pe_circuit` and `unitary` must be passed.
+            state_preparation: The circuit that prepares the state whose eigenphase will be
+                                 measured.  If this parameter is omitted, no preparation circuit
+                                 will be run and input state will be the all-zero state in the
+                                 computational basis.
+            pe_circuit: The phase estimation circuit.
+            num_unitary_qubits: Must agree with the number of qubits in the unitary in `pe_circuit`
+                                if `pe_circuit` is passed. This parameter will be set from `unitary`
+                                if `unitary` is passed.
+            quantum_instance: The quantum instance on which the circuit will be run.
+
+           Raises:
+               ValueError: unless only one of `unitary` and `pe_circuit` is `None`.
+                           `num_unitary_qubits` disagrees with size of `unitary`.
+        """
+
+        # Determine if user passed a unitary, or the entire QPE circuit with the unitary
+        # already embedded, and set properties.
+        if unitary is None:
+            if pe_circuit is None:
+                raise ValueError('Only one of `unitary` and `pe_circuit` may be `None`.')
+
+            if num_unitary_qubits is None:
+                raise ValueError('`num_unitary_qubits` is required when passing `pe_circuit`.')
+            self._pe_circuit = pe_circuit
+            self._num_unitary_qubits = num_unitary_qubits
+        else:
+            if not (num_unitary_qubits is None or num_unitary_qubits == unitary.num_qubits):
+                raise ValueError('`num_unitary_qubits` disagrees with size of `unitary`.')
+            pe_circuit = PhaseEstimation(num_evaluation_qubits, unitary)
```

Comment:
This could be put straight into self._pre_circuit instead of overwriting the supplied param. Then below if state_prep is not None self._pe_circuit = self._pe_circuit.compose(... And the else part seems like its unneccessary since self._pe_circuit would already have pe_circuit in it anyway from above, i.e. it only needed updating if there was some state prep to add.